[Docs] Add alerts cooldown period feature (#9739)
mlrun/mlrun: docs/concepts/alerts.md

- `alerts.mode` &mdash; Enables/disables the feature. Enabled by default.
- `alerts.max_allowed` &mdash; Maximum number of alerts allowed to be configured, by default 10000. Any new alerts above this limit return an error.
- `alerts.max_criteria_count` &mdash; Maximum number of events. By default, 100.

> - `alerts.cooldown_reset_interval` &mdash; Interval at which alerts, whose `cooldown_period` have completed, are reset. By default, 15s.

Reviewer: ```suggestion
- `alerts.cooldown_reset_interval` &mdash; Interval at which alerts, whose `cooldown_period` have completed, are reset. By default, 15 seconds.
```

Author: updated

- {py:func}`~mlrun.projects.MlrunProject.delete_alert_config` &mdash; Delete an alert
- {py:func}`~mlrun.projects.MlrunProject.get_alert_template` &mdash; Retrieve a specific alert template
- {py:func}`~mlrun.projects.MlrunProject.list_alert_templates` &mdash; Retrieve the list of all alert templates
- {py:func}`~mlrun.projects.MlrunProject.list_alerts_configs` &mdash; Retrieve the list of alerts of a project

## Predefined events (`EventKind`)
The predefined event types ({py:class}`~mlrun.common.schemas.alert.EventKind`) are:
- `concept-drift-detected` &mdash; A detected change, over time, of  statistical properties of the target variable (what the model is predicting). 
- `concept-drift-suspected` &mdash; A suspected change, over time, of  statistical properties of the target variable (what the model is predicting). 
- `data-drift-detected` &mdash; A detected change in model input data that potentially leads to model performance degradation. 
- `data-drift-suspected` &mdash; A suspected change in model input data that potentially leads to model performance degradation. 
- `mm-app-anomaly-detected` &mdash; An alert based on user-defined metrics/results.

> - `mm-app-anomaly-suspected` &mdash; An alert based on user-defined metrics/results.

Reviewer: identical explanations ?

Author: updated

- `mm-app-failed` &mdash; A model monitoring app failed.

> - `model-monitoring-lag-detected` &mdash; Rhe monitoring writer falls behind the processing of inference events.

Reviewer: ```suggestion
- `model-monitoring-lag-detected` &mdash; The monitoring writer falls behind the processing of inference events.
```

- `failed` &mdash; The job failed.

See {ref}`model-monitoring-overview` for more details on drift and performance.

## Create an alert
When creating an alert you can select an event type for a specific model, for example `data_drift_suspected` or any of the predefined events above.
You can optionally specify the frequency of the alert using the criteria field, which controls the threshold number of events in a given time window that triggers the alert.
If criteria is not specified, the default is `count=1` and `period=None`, in which case the alert triggers immediately upon the first matching event.

You can configure Slack, Git, and webhook {ref}`notifications` for alerts.

``` {Admonition} Note on run identification
Alerts track the job runs by name (`run.metadata.name`), not by the unique run UID. The run name can either be set explicitly or automatically generated when a job is executed. 
You can access the run name from the result of the `run_function` call, for example:
```python
run = project.run_function("my-function", handler="handler", local=True)
run_id = run.metadata.name
```
See all of the {py:class}`~mlrun.alerts.alert.AlertConfig` parameters. 

> ### Cooldown period

Reviewer 2: I would add to the description the motivation for using the cooldown period: Enable the user to choose to receive alerts at a reasonable frequency, so on the one hand, the user will keep receiving alerts in case the issue happens again, but on the other hand, those alerts won't be one right after the other, which is the case when no cooldown period is defined

Author: updated

The `cooldown_period` parameter of `AlertConfig` can be used to delay resetting an alert. 

> The alert remains active for the duration of the `cooldown_period`, and incoming events are ignored. 

Reviewer: ```suggestion
The alert remains active for the duration of the `cooldown_period`, and incoming events are ignored (no additional notifications are sent). 
```

After the cooldown period expires and the alert is reset, it can be triggered again by new events.
The reset timing is approximate: reset happens when a periodic task runs, by default, every ~15s. 
By default, the `cooldown_period` is not set.

Guidelines:
- The `reset_policy` must be set to `auto`.

> - The `cooldown_period` must be >0. (If set to 0, alarms are reset immediately.)

Reviewer: ```suggestion
- The `cooldown_period` must be >0. (If set to 0, alerts are reset immediately.)
```

> - The `cooldown_period` must be >0. (If set to 0, alarms are reset immediately.)

Reviewer: This bullet contradicts itself - "must be >0" followed by "If set to 0"... cant both be true
 A positive `cooldown_period` is only allowed when `reset_policy=auto`, and it must be at least `cooldown_reset_interval`

Author: That is an explanation, not a contradiction. Similar to the docstring

- Cooldown periods can be set as, for example, 1d, 3h, 5m, 15s, etc.
- `cooldown_period` must be at least `cooldown_reset_interval`. See [System configuration](#system-configuration).

> You can manually reset an alarm at any time ({py:func}`~mlrun.projects.MlrunProject.reset_alert_config`), whether or not the `cooldown_period` is active.

Reviewer: ```suggestion
You can manually reset an alert at any time ({py:func}`~mlrun.projects.MlrunProject.reset_alert_config`), whether or not the `cooldown_period` is active.
```

Author: updated

- {py:func}`~mlrun.projects.MlrunProject.reset_alert_config` &mdash; Reset an alert
- {py:func}`~mlrun.projects.MlrunProject.delete_alert_config` &mdash; Delete an alert
- {py:func}`~mlrun.projects.MlrunProject.get_alert_template` &mdash; Retrieve a specific alert template
- {py:func}`~mlrun.projects.MlrunProject.list_alert_templates` &mdash; Retrieve the list of all alert templates
- {py:func}`~mlrun.projects.MlrunProject.list_alerts_configs` &mdash; Retrieve the list of alerts of a project

## Predefined events (`EventKind`)
The predefined event types ({py:class}`~mlrun.common.schemas.alert.EventKind`) are:
- `concept-drift-detected` &mdash; A detected change, over time, of  statistical properties of the target variable (what the model is predicting). 
- `concept-drift-suspected` &mdash; A suspected change, over time, of  statistical properties of the target variable (what the model is predicting). 
- `data-drift-detected` &mdash; A detected change in model input data that potentially leads to model performance degradation. 
- `data-drift-suspected` &mdash; A suspected change in model input data that potentially leads to model performance degradation. 

> - `mm-app-anomaly-detected` &mdash; A detected alert based on user-defined metrics/results.

Reviewer: an alert isn't itself "detected/suspected", consider :

```suggestion
- `mm-app-anomaly-detected` &mdash; An alert triggered by a detected anomaly in user-defined metrics/results.
```
also to the next point

> - `mm-app-anomaly-suspected` &mdash; A suspoected alert based on user-defined metrics/results.

Reviewer: ```suggestion
- `mm-app-anomaly-suspected` &mdash; A suspected alert based on user-defined metrics/results.
```

> - `mm-app-anomaly-suspected` &mdash; A suspected alert based on user-defined metrics/results.

Reviewer: do the same change as in the previous one

You can configure Slack, Git, and webhook {ref}`notifications` for alerts.

``` {Admonition} Note on run identification
Alerts track the job runs by name (`run.metadata.name`), not by the unique run UID. The run name can either be set explicitly or automatically generated when a job is executed. 
You can access the run name from the result of the `run_function` call, for example:
```python
run = project.run_function("my-function", handler="handler", local=True)
run_id = run.metadata.name
```
See all of the {py:class}`~mlrun.alerts.alert.AlertConfig` parameters. 

### Cooldown period

The `cooldown_period` parameter of `AlertConfig` can be used to delay resetting an alert. 
When the `cooldown_period` is active, the alert remains active for its duration, but incoming events are ignored:
**you do not receive continuous alerts for an ongoing situation**.
After the cooldown period expires and the alert is reset, it can be triggered again by new events.
The reset timing is approximate: reset happens when a periodic task runs, by default, every ~15s. 
By default, the `cooldown_period` is not set.

Guidelines:
- The `reset_policy` must be set to `auto`.

> - The `cooldown_period` must be >0 and >`cooldown_reset_interval`. See [System configuration](#system-configuration). (When `cooldown_period` is not set or set to 0, alerts are reset immediately.)

Reviewer: ```suggestion
- The `cooldown_period` must be >0 and >=`cooldown_reset_interval`. See [System configuration](#system-configuration). (When `cooldown_period` is not set or set to 0, alerts are reset immediately.)
```

    ),
    trigger=alert_objects.AlertTrigger(events=[event_name]),
    criteria=alert_objects.AlertCriteria(period="10m", count=3),

>     cooldown_period=30s,

Reviewer: ```suggestion
    cooldown_period="30s",
```
